File: Python/11-Math&Geometry/Algorithms/2-Sieve-of-Eratosthenes(PrimeNumbers)/Sieve-of-Eratosthenes(PrimeNumbers)Pro.py
import math
import numpy as np
from itertools import compress
from multiprocessing import Pool

def sieve_basic(n):
    """Basic Sieve of Eratosthenes"""
    primes = [True] * (n + 1)
    primes[0], primes[1] = False, False
    for i in range(2, int(n**0.5) + 1):
        if primes[i]:
            for j in range(i * i, n + 1, i):
                primes[j] = False
    return [i for i, is_prime in enumerate(primes) if is_prime]


def sieve_optimized(n):
    """Optimized Sieve: Skips even numbers after 2"""
    if n < 2:
        return []
    primes = [True] * (n + 1)
    primes[0], primes[1] = False, False
    for i in range(4, n + 1, 2):
        primes[i] = False
    for i in range(3, int(n**0.5) + 1, 2):
        if primes[i]:
            primes[i * i: n + 1: i * 2] = [False] * len(range(i * i, n + 1, i * 2))
    return [i for i, is_prime in enumerate(primes) if is_prime]


def sieve_numpy(n):
    """Using NumPy for efficient array processing"""
    if n < 2:
        return []
    primes = np.ones(n + 1, dtype=bool)
    primes[:2] = False
    for i in range(2, int(n**0.5) + 1):
        if primes[i]:
            primes[i * i:n + 1:i] = False
    return np.nonzero(primes)[0].tolist()


def sieve_bitset(n):
    """Using a boolean bitset approach with itertools.compress"""
    if n < 2:
        return []
    primes = [True] * (n + 1)
    primes[0], primes[1] = False, False
    for i in range(2, int(math.sqrt(n)) + 1):
        if primes[i]:
            primes[i * i:n + 1:i] = [False] * len(range(i * i, n + 1, i))
    return list(compress(range(n + 1), primes))


def segmented_sieve(limit):
    """Segmented Sieve of Eratosthenes for large numbers"""
    prime_list = sieve_basic(int(math.sqrt(limit)))
    primes = []
    low, high = 2, limit
    segment = [True] * (high - low + 1)
    for p in prime_list:
        start = max(p * p, (low + p - 1) // p * p)
        for j in range(start, high + 1, p):
            segment[j - low] = False
    primes.extend([low + i for i in range(high - low + 1) if segment[i]])
    return primes


def mark_non_primes(start, n, base_primes):
    """Marks non-primes in a segment starting from 'start'."""
    segment = [True] * ((n - start) + 1)
    for p in base_primes:
        begin = max(p * p, (start + p - 1) // p * p)
        for j in range(begin, n + 1, p):
            segment[j - start] = False
    return [start + i for i, is_prime in enumerate(segment) if is_prime]


def parallel_sieve(n):
    """Parallelized sieve using multiprocessing"""
    base_primes = sieve_basic(int(math.sqrt(n)))

    step = n // 4  # Divide work into 4 segments
    ranges = [(i, min(i + step, n)) for i in range(2, n, step)]

    with Pool() as pool:
        result = pool.starmap(mark_non_primes, [(r[0], r[1], base_primes) for r in ranges])

    return [2] + sorted(set(base_primes + [num for sublist in result for num in sublist]))


# Display outputs for a sample range
n = 100
print("Basic Sieve:", sieve_basic(n))
print("Optimized Sieve:", sieve_optimized(n))
print("NumPy Sieve:", sieve_numpy(n))
print("Bitset Sieve:", sieve_bitset(n))
print("Segmented Sieve:", segmented_sieve(n))
print("Parallel Sieve:", parallel_sieve(n))

File: JavaScript/11-Math&Geometry/Algorithms/2-Sieve-of-Eratosthenes(PrimeNumbers)/Sieve-of-Eratosthenes(PrimeNumbers)Pro.js
/**
 * Sieve of Eratosthenes - Five Implementations
 * ----------------------------------------
 *
 * This code demonstrates five different implementations of the Sieve of Eratosthenes algorithm
 * for finding prime numbers up to a given limit. Each implementation is tailored to a
 * specific real-world project scenario.
 */

/**
 * 1. Basic Sieve of Eratosthenes
 *
 * - Description:  The most straightforward implementation.  Good for understanding
 * the core algorithm.
 * - Application:  Educational tool for demonstrating prime number generation.
 */
function basicSieve(limit) {
    if (limit <= 1) return [];

    const isPrime = new Array(limit + 1).fill(true);
    isPrime[0] = isPrime[1] = false;

    for (let p = 2; p * p <= limit; p++) {
        if (isPrime[p]) {
            for (let i = p * p; i <= limit; i += p) {
                isPrime[i] = false;
            }
        }
    }

    const primes = [];
    for (let i = 2; i <= limit; i++) {
        if (isPrime[i]) {
            primes.push(i);
        }
    }
    return primes;
}

// Example Usage:
const basicPrimes = basicSieve(100);
console.log("1. Basic Sieve (Educational Tool):", basicPrimes);


/**
 * 2. Sieve with Bit Array Optimization
 *
 * - Description:  Uses a bit array to reduce memory usage.  Each bit represents a number.
 * - Application:  Finding primes in embedded systems with limited memory.
 */
function sieveWithBitArray(limit) {
    if (limit <= 1) return [];

    const numBits = Math.ceil((limit + 1) / 32);
    const bitArray = new Uint32Array(numBits);

    // Helper functions to set and check bits
    const setBit = (arr, n) => {
        const wordIndex = Math.floor(n / 32);
        const bitIndex = n % 32;
        arr[wordIndex] |= (1 << bitIndex);
    };

    const checkBit = (arr, n) => {
        const wordIndex = Math.floor(n / 32);
        const bitIndex = n % 32;
        return (arr[wordIndex] & (1 << bitIndex)) !== 0;
    };

    // Initialize all bits to 1 (prime)
    for (let i = 0; i < numBits; i++) {
        bitArray[i] = 0xFFFFFFFF; // Set all 32 bits to 1
    }
    setBit(bitArray, 0); // 0 is not prime
    setBit(bitArray, 1); // 1 is not prime

    for (let p = 2; p * p <= limit; p++) {
        if (checkBit(bitArray, p)) {
            for (let i = p * p; i <= limit; i += p) {
                setBit(bitArray, i);
            }
        }
    }

    const primes = [];
    for (let i = 2; i <= limit; i++) {
        if (checkBit(bitArray, i)) {
            primes.push(i);
        }
    }
    return primes;
}

// Example Usage:
const bitArrayPrimes = sieveWithBitArray(100);
console.log("2. Sieve with Bit Array (Embedded System):", bitArrayPrimes);


/**
 * 3. Segmented Sieve
 *
 * - Description: Divides the range into smaller segments to reduce memory usage.
 * - Application: Finding primes in very large ranges, such as in cryptographic applications.
 */
function segmentedSieve(limit) {
    if (limit <= 1) return [];
    const segmentSize = Math.floor(Math.sqrt(limit));
    const primes = [];
    const firstSegmentPrimes = basicSieve(segmentSize); // Use basic sieve for first segment

    for (let i = 0; i * segmentSize <= limit; i++) {
        const low = i * segmentSize;
        const high = Math.min((i + 1) * segmentSize - 1, limit);
        const isPrime = new Array(high - low + 1).fill(true);

        for (const p of firstSegmentPrimes) {
            let start = Math.max(p * p, Math.ceil(low / p) * p);
            for (let j = start - low; j <= high - low; j += p) {
                isPrime[j] = false;
            }
        }

        for (let j = low; j <= high; j++) {
            if (isPrime[j - low]) {
                primes.push(j);
            }
        }
    }
    return primes;
}

// Example Usage:
const segmentedPrimes = segmentedSieve(1000);
console.log("3. Segmented Sieve (Large Range):", segmentedPrimes.slice(0, 50), "..."); // Print only the first 50


/**
 * 4. Sieve with Wheel Factorization (Wheel 6)
 *
 * - Description: Uses a "wheel" to skip multiples of small primes (2, 3) to improve efficiency.
 * - Application:  Optimized prime generation for number theory research.
 */
function sieveWithWheel6(limit) {
    if (limit <= 1) return [];

    const primes = [2, 3]; // Initial primes
    const isPrime = new Array(Math.floor((limit - 1) / 6) + 1).fill(true);

    for (let i = 5, w = 2; i * i <= limit; w = 6 - w, i += w) {
        if (isPrime[Math.floor((i - 1) / 6)]) {
            for (let j = i * i, step = w; j <= limit; step = 6 - step, j += step * i) {
                isPrime[Math.floor((j - 1) / 6)] = false;
            }
        }
    }

    for (let i = 5, w = 2; i <= limit; w = 6 - w, i += w) {
        if (isPrime[Math.floor((i - 1) / 6)]) {
            primes.push(i);
        }
    }
    return primes;
}

// Example Usage:
const wheel6Primes = sieveWithWheel6(200);
console.log("4. Sieve with Wheel 6 (Number Theory):", wheel6Primes);


/**
 * 5. Sieve with Incremental Sieving
 *
 * - Description:  Generates primes in chunks and combines the results. Useful for
 * applications where primes are needed in a streaming fashion.
 * - Application:  Generating prime numbers for a real-time game or simulation.
 */
function incrementalSieve(limit, callback) {
     if (limit <= 1) return [];

    const segmentSize = 1000; // Tune this based on performance
    let currentStart = 2;

    while (currentStart <= limit) {
        const currentEnd = Math.min(currentStart + segmentSize - 1, limit);
        const isPrime = new Array(currentEnd - currentStart + 1).fill(true);

        for (let p = 2; p * p <= currentEnd; p++) {
            if(p >= currentStart){
               let start = p * p;
               for (let i = start - currentStart; i <= currentEnd - currentStart; i += p) {
                    isPrime[i] = false;
                }
            }
            else{
                let start = Math.ceil(currentStart / p) * p;
                 for (let i = start - currentStart; i <= currentEnd - currentStart; i += p) {
                    isPrime[i] = false;
                }
            }

        }

        for (let i = currentStart; i <= currentEnd; i++) {
            if (isPrime[i - currentStart]) {
                callback(i); // Send prime to callback
            }
        }
        currentStart = currentEnd + 1;
    }
}
// Example Usage:
const incrementalPrimes = [];
incrementalSieve(150, (prime) => {
    incrementalPrimes.push(prime);
});
console.log("5. Incremental Sieve (Real-time Game):", incrementalPrimes);
